spec/Section 3 -- Type System.md

```diff
@@ -263,74 +263,108 @@ TypeDefinition :
 
-GraphQL supports two abstract types: interfaces and unions.
+### Leaf and Composite Types
 
-An `Interface` defines a list of fields; `Object` types and other Interface
-types which implement this Interface are guaranteed to implement those fields.
-Whenever a field claims it will return an Interface type, it will return a
-valid implementing Object type during execution.
+:: A *leaf type* is found at the leaves in a response tree. There are two kinds
+of leaf types: *Scalar type* and *Enum type*.
 
-A `Union` defines a list of possible types; similar to interfaces, whenever the
-type system claims a union will be returned, one of the possible types will be
-returned.
+The most basic *leaf type* is the *Scalar type*. A scalar represents a primitive
+value, like a string or an integer. GraphQL defines a set of *built-in scalars*
+and also allows a service to additionally provide any *custom scalar*.
 
-Finally, oftentimes it is useful to provide complex structs as inputs to
-GraphQL field arguments or variables; the `Input Object` type allows the schema
-to define exactly what data is expected.
+Oftentimes, the possible values for a *leaf type* are enumerable. GraphQL offers
+the *Enum type* in those cases, where the each enum type definition specifies
+the set of possible values.
 
+:: A *composite type* compose together leaf types and other composite types as a
+set of named fields, allowing the composition of arbitrary type hierarchies and
+nested trees of data. There are three composite *output type*: Object,
+Interface, and Union, and one composite *input type*: Input Object.
 
-### Wrapping Types
+The *Object type* is the concrete output *composite type* which describes an
+actual concrete value which will a field will resolve during execution, composed
+of other *output type* fields, allowing responses of nested data.
+
+Oftentimes it is useful to provide a *composite type* as an *input type* to a
+field argument or variable; the *Input Object type* composes a set of
+*input type* fields, allowing inputs of nested data.
+
+
+### Abstract Types
+
+:: An *abstract type* defines a set of possible types, one of which will be
+resolved during field execution. There are two kinds of abstract types:
+*Interface type* and *Union type*.
+
+:: The type an abstract type resolves to during execution is referred to as the
+*concrete Object type*, which is always an *Object type*.
+
+An *Interface type* defines a list of fields. An type which implements this
+Interface must also implement those fields. Whenever a field claims it will
+return an *Interface type*, it will return an implementing
+*concrete Object type* during field execution.
+
+A *Union type* defines a list of possible types. Similar to Interfaces,
+whenever a field claims it will return a *Union type*, it will return one of the
+possible types as the *concrete Object type* during field execution.
+
+
+### Wrapped Types
```

Comment:
@benjie - would love your thoughts on this newly written subsection.

This is based on [your comment](https://github.com/graphql/graphql-spec/pull/845#issuecomment-826341088) with very minor changes:

* **Wrapped type** - the kind of type including Lists and Non-Nulls
  * **List type** - changed from your suggestion of **List wrapped type**
  * **Non-Null type** - changed from your suggestion of **Non-Null wrapped type**
* **Inner type** - the type immediately wrapped (eg the `ofType`)
  * **Item type** - I added this - the inner type of a List
  * **Nullable type** - I added this - the inner type of a Non-Null